In [ ]:
!pip install -q pyomo

     |████████████████████████████████| 9.7 MB 5.3 MB/s 
     |████████████████████████████████| 49 kB 5.7 MB/s 


In [ ]:
from pyomo.environ  import *

In [ ]:
import numpy as np

In [ ]:
coef = np.loadtxt('lab4_ex1_coef.txt', delimiter=',')

In [ ]:
print(coef)

[[  0.65   0.75  -5.85  -8.35 -12.38  30.     0.  ]
 [ -0.5    0.    -3.8   -0.9    0.9    0.    24.8 ]
 [  1.     0.     0.    -1.     0.    -1.     0.8 ]
 [  0.    -2.9   -1.9    0.     0.    -0.9  -71.4 ]
 [  0.     0.     0.     4.7    8.02   6.15  95.  ]
 [ -0.7   -1.4    7.9   -2.9    0.9    0.   -47.  ]]


In [ ]:
print(coef.shape)

(6, 7)


In [ ]:
N = coef.shape[1]-1 #no of variables
M = coef.shape[0]-1 #no of constraints

In [ ]:
col_indices = np.arange(N)
row_indices = np.arange(M)

In [ ]:
obj_coef_c = coef[0,:-1]
cons_coef_A = coef[1:,:-1]
cons_rhs_b = coef[1:,-1]

In [ ]:
lb = 0
ub = np.inf

In [ ]:
model_ex1 = ConcreteModel()

In [ ]:
model_ex1.x = Var(col_indices)

In [ ]:
model_ex1.obj = Objective(expr = summation(obj_coef_c,model_ex1.x),sense = maximize)

In [ ]:
model_ex1.constr = ConstraintList()

In [ ]:
for i in row_indices:
  model_ex1.constr.add(expr = summation(cons_coef_A[i],model_ex1.x) <= cons_rhs_b[i])

In [ ]:
for i in col_indices:
  model_ex1.x[i].setlb(lb)
  model_ex1.x[i].setub(ub)

In [ ]:
model_ex1.pprint()

2 Set Declarations
    constr_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {0, 1, 2, 3, 4, 5}

1 Var Declarations
    x : Size=6, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True :  Reals
          1 :     0 :  None :  None : False :  True :  Reals
          2 :     0 :  None :  None : False :  True :  Reals
          3 :     0 :  None :  None : False :  True :  Reals
          4 :     0 :  None :  None : False :  True :  Reals
          5 :     0 :  None :  None : False :  True :  Reals

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 0.65*x[0] + 0.75*x[1] - 5.85*x[2] - 8.35*x[

In [ ]:
!apt-get install -y -qq coinor-cbc

Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 155676 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to unpack .../4-coinor-libcbc3_2.9.9+repack1-1_amd64.deb ...
Unpacking coinor-libcbc3 (2.9.9+repack1-1) ...
Selecting p

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

  - termination condition: unbounded
  - message from solver: <undefined>


Solver status: warning
Solver termination condition: unbounded


$\large{\textbf{Answer 1.7}} \\ \text{The solver status is displaying WARNING which may be because of the termination condition that is coming out to be UNBOUNDED.} \\ \text{A possible reason for the system to be unbounded is that maybe the sense of the objective function is wrong or the constraints are not able to bound the region.}$

In [ ]:
model_ex1.obj.set_sense(minimize)

In [ ]:
model_ex1.pprint()

2 Set Declarations
    constr_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {0, 1, 2, 3, 4, 5}

1 Var Declarations
    x : Size=6, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True :  Reals
          1 :     0 :  None :  None : False :  True :  Reals
          2 :     0 :  None :  None : False :  True :  Reals
          3 :     0 :  None :  None : False :  True :  Reals
          4 :     0 :  None :  None : False :  True :  Reals
          5 :     0 :  None :  None : False :  True :  Reals

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 0.65*x[0] + 0.75*x[1] - 5.85*x[2] - 8.35*x[

In [ ]:
result = opt_cbc.solve(model_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

  - termination condition: unbounded
  - message from solver: <undefined>


Solver status: warning
Solver termination condition: unbounded


$\text{Even after changing the sense of the objective function, we observe that the termination condition is still coming out to be unbounded,} \\ \text{so we can say that the sense of the objective function is correct but there is some issue in one of the constraints or the bounds of some variable.}$

In [ ]:
model_ex1.obj.set_sense(maximize)

In [ ]:
model_ex1.x[1].setub(1200)

In [ ]:
result = opt_cbc.solve(model_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
print('\nObjective = ', model_ex1.obj())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model_ex1.x[i].value)
print('\nConstraints')
model_ex1.constr.display()


Objective =  1373.9752701

Decision Variables
x[ 0 ] =  16.247154
x[ 1 ] =  1200.0
x[ 2 ] =  0.0
x[ 3 ] =  0.0
x[ 4 ] =  0.0
x[ 5 ] =  15.447154

Constraints
constr : Size=5
    Key : Lower : Body               : Upper
      1 :  None :          -8.123577 :  24.8
      2 :  None : 0.7999999999999989 :   0.8
      3 :  None :      -3493.9024386 : -71.4
      4 :  None :         94.9999971 :  95.0
      5 :  None :      -1691.3730078 : -47.0


$\large{\textbf{Answer 1.9 :}} \\ \text{Here we can see that we get an optimal solution as soon as we set an upper bound for } x_2 \\ \text{The possible reason for that might be in the constraints we observe that if we keep any} \\ \text{four variables constant we get some upper bounds for the fifth one other than  } x_2 \\ \text{When there is no upper bound for }x_2, \text{due to the sense of maximisation, the value of the objective function comes out to be } + ∞$

$\text{Upper bound of } x_2 = 1200 \\ \text{Objective Function Value = 1373.975} \\ \textbf{Value of decision variables:} \\ x_1 = 16.247\\ x_2 = 1200 \\ x_3 = 0\\ x_4 = 0 \\ x_5 = 0 \\ x_6 = 15.447 \\ \textbf{Constraint activities:} \\ \textbf{Constaint 1:} \text{ It is inactive as its value is coming out to be -8.123 which is strictly less than the RHS = 24.8 } \\ \textbf{Constaint 2:} \text{ It is active as its value is coming out to be 0.8 which is exactly equal to the RHS.} \\ \textbf{Constaint 3:} \text{ It is inactive as its value is coming out to be -3493.9024 which is strictly less than the RHS = -71.4} \\ \textbf{Constaint 4:} \text{ It is active as its value is coming out to be 95 which is exactly equal to the RHS.} \\ \textbf{Constaint 5:} \text{ It is inactive as its value is coming out to be -1691.373 which is strictly less than the RHS = -47 } \\ \text{Here we can also observe that } x_2 \text{ is attaining the upper bound which verifies our claim that } x_2 \text{ was tending to } ∞ \text{ thus making the solution unbounded.}$